sdk/identity/azure-identity/azure/identity/_credentials/app_service.py

```diff
@@ -0,0 +1,113 @@
+# ------------------------------------
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+# ------------------------------------
+import functools
+import os
+from typing import TYPE_CHECKING
+
+from azure.core.credentials import AccessToken
+from azure.core.pipeline.transport import HttpRequest
+
+from .. import CredentialUnavailableError
+from .._constants import EnvironmentVariables
+from .._internal.managed_identity_client import ManagedIdentityClient
+from .._internal.get_token_mixin import GetTokenMixin
+
+if TYPE_CHECKING:
+    from typing import Any, Optional
+
+
+class AppServiceCredential(GetTokenMixin):
+    def __init__(self, **kwargs):
+        # type: (**Any) -> None
+        super(AppServiceCredential, self).__init__()
+
+        client_args = _get_client_args(**kwargs)
+        if client_args:
+            self._client = ManagedIdentityClient(**client_args)
+        else:
+            self._client = None
+
+    def get_token(self, *scopes, **kwargs):
+        # type: (*str, **Any) -> AccessToken
+        if not self._client:
+            raise CredentialUnavailableError(
+                message="App Service managed identity configuration not found in environment"
+            )
+        return super(AppServiceCredential, self).get_token(*scopes, **kwargs)
+
+    def _acquire_token_silently(self, *scopes):
+        # type: (*str) -> Optional[AccessToken]
+        return self._client.get_cached_token(*scopes)
+
+    def _request_token(self, *scopes, **kwargs):
+        # type: (*str, **Any) -> AccessToken
+        return self._client.request_token(*scopes, **kwargs)
+
+
+def _get_client_args(**kwargs):
+    # type: (dict) -> Optional[dict]
+    identity_config = kwargs.pop("identity_config", None) or {}
+
+    url = os.environ.get(EnvironmentVariables.IDENTITY_ENDPOINT)
+    secret = os.environ.get(EnvironmentVariables.IDENTITY_HEADER)
+    if url and secret:
+        version = "2019-08-01"
+        base_headers = {"X-IDENTITY-HEADER": secret}
+        content_callback = None
+    else:
+        url = os.environ.get(EnvironmentVariables.MSI_ENDPOINT)
+        secret = os.environ.get(EnvironmentVariables.MSI_SECRET)
+        if not (url and secret):
+            # App Service managed identity isn't available in this environment
+            return None
+
+        version = "2017-09-01"
+        base_headers = {"secret": secret}
+        content_callback = _parse_app_service_expires_on
+        if kwargs.get("client_id"):
+            identity_config["clientid"] = kwargs.pop("client_id")
```

Comment:
If "client_id" does not exist, I guess we don't want to set "clientid" to "None"?

Reply:
Nevermind. :)